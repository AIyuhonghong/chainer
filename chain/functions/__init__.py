# Non-parameterized functions
from accuracy import accuracy
from dropout  import dropout
from identity import identity
from relu     import relu
from sigmoid  import sigmoid
from softmax  import softmax
from softmax_cross_entropy import softmax_cross_entropy
from tanh     import tanh

# Parameterized layers
from embed_id import EmbedID
from linear   import Linear
