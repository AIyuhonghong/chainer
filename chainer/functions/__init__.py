# Non-parameterized functions
from accuracy   import accuracy
from concat     import concat
from dropout    import dropout
from identity   import identity
from lstm       import lstm
from pooling_2d import average_pooling_2d, max_pooling_2d
from relu       import relu
from sigmoid    import sigmoid
from softmax    import softmax
from softmax_cross_entropy import softmax_cross_entropy
from sum        import sum
from tanh       import tanh

# Parameterized layers
from convolution_2d import Convolution2D
from embed_id       import EmbedID
from inception      import Inception
from linear         import Linear
