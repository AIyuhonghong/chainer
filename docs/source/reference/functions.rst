Standard Function implementations
=================================

.. currentmodule:: chainer.functions

Parameterized function classes
------------------------------
.. autoclass:: BatchNormalization
.. autoclass:: Convolution2D
.. autoclass:: EmbedID
.. autoclass:: Linear
.. autoclass:: Parameter
.. autoclass:: PReLU

Array manipulation functions
----------------------------
.. autofunction:: concat
.. autofunction:: copy
.. autofunction:: dropout
.. autofunction:: identity

Activation functions
--------------------
.. autofunction:: leaky_relu
.. autofunction:: lstm
.. autofunction:: relu
.. autofunction:: sigmoid
.. autofunction:: softmax
.. autofunction:: tanh

Pooling functions
-----------------
.. autofunction:: average_pooling_2d
.. autofunction:: max_pooling_2d

Loss, evaluation and aggregation
--------------------------------
.. autofunction:: accuracy
.. autofunction:: mean_squared_error
.. autofunction:: softmax_cross_entropy
.. autofunction:: sum
