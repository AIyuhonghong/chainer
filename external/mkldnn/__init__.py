from external.mkldnn import prepare_mkldnn  # NOQA
from external.mkldnn.prepare_mkldnn import prepare  # NOQA
from external.mkldnn.prepare_mkldnn import root  # NOQA
from external.mkldnn.prepare_mkldnn import lib_path  # NOQA
from external.mkldnn.prepare_mkldnn import include_path  # NOQA
from external.mkldnn.prepare_mkldnn import source_path  # NOQA
