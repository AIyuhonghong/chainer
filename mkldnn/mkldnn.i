%module mkldnn

%{
    #define SWIG_FILE_WITH_INIT
    #include "common.h"
    #include "convolution.h"
    #include "stream_factory.h"
%}

%include "numpy.i"

%init %{
    import_array();
    global_init();
%}

%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* x, int x_d1, int x_d2, int x_d3, int x_d4 )}
%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* gx, int gx_d1, int gx_d2, int gx_d3, int gx_d4 )}
%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* W, int W_d1, int W_d2, int W_d3, int W_d4 )}
%apply ( float* IN_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* gW, int gW_d1, int gW_d2, int gW_d3, int gW_d4 )}
%apply ( float* IN_ARRAY1, int DIM1)
    {( float* b, int b_d1)}
%apply ( float* IN_ARRAY1, int DIM1)
    {( float* gb, int gb_d1)}
%apply ( float* INPLACE_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* y, int y_d1, int y_d2, int y_d3, int y_d4 )}
%apply ( float* INPLACE_ARRAY4, int DIM1, int DIM2, int DIM3, int DIM4 )
    {( float* gy, int gy_d1, int gy_d2, int gy_d3, int gy_d4 )}

%include "common.h"
%include "convolution.h"
%include "stream_factory.h"

%template(Convolution2D_F32) Convolution2D<float>;
