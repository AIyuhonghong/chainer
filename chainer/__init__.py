import pkg_resources

from chainer import function
from chainer import function_set
from chainer.functions import basic_math
from chainer import optimizer
from chainer import serializer
from chainer import variable


__version__ = pkg_resources.get_distribution('chainer').version

AbstractSerializer = chainer.AbstractSerializer
Deserializer = chainer.Deserializer
Function = function.Function
FunctionSet = function_set.FunctionSet
Optimizer = optimizer.Optimizer
Serializer = chainer.Serializer
Variable = variable.Variable

basic_math.install_variable_arithmetics()
