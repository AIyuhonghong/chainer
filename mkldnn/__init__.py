import sys
import ctypes

# For C++ extension to work
sys.setdlopenflags(sys.getdlopenflags() | ctypes.RTLD_GLOBAL)

# API lift
from mkldnn import api
from mkldnn import chainer
from mkldnn import compute_complex

from mkldnn.mdarray import mdarray

from mkldnn.chainer.fanout import *

from mkldnn.config import global_config  # NOQA
global_config.gx_opt = True
