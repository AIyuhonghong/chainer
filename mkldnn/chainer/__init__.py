import os
import numpy

from mkldnn.mdarray import mdarray

from chainer import variable
from chainer import configuration

from chainer.configuration import config
from chainer.configuration import global_config

# export CHAINER_ENABLE_COSIM=0
global_config.cosim = bool(int(os.environ.get('CHAINER_ENABLE_COSIM', '0')))
# export CHAINER_ENABLE_COSIM_CONTINUE=0
global_config.cosim_continue = bool(int(os.environ.get('CHAINER_ENABLE_COSIM_CONTINUE', '0')))

if global_config.cosim is True:
    variable.RANK_START = 1


def is_cosim_continue():
    """Get the cosim continue mode

    Returns:
        bool: Return ``True`` if continue after a failure.
    """
    return config.cosim_continue


def is_cosim():
    """Get the cosim mode.

    Returns:
        bool: Return ``True`` if Chainer is in cosim mode.
    """
    return config.cosim


def plain_array(params):
    assert isinstance(params, tuple) \
            or isinstance(params, list) \
            or isinstance(params, mdarray) \
            or isinstance(params, numpy.ndarray) \
            or isinstance(params, variable.Variable)

    _params = ()

    if isinstance(params, variable.Variable):
        return (numpy.array(params.data), )
    elif isinstance(params, numpy.ndarray):
        return (params, )
    elif isinstance(params, mdarray):
        return (numpy.array(params), )

    for p in params:
        if isinstance(p, variable.Variable):
            p = numpy.array(p.data)
        if isinstance(p, mdarray):
            _params += (numpy.array(p), )
        else:
            _params += (p, )

    return _params

