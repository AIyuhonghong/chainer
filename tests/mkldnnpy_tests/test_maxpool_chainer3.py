import chainer
import chainer.functions as F
import chainer.links as L
import numpy as np
import math
from mkldnn import switch

#input = [   [ 1, 2, 3 ],
#            [ 4, 5, 6 ],
#            [ 7, 8, 9 ] ]

(n,c,h,w)=(2,3,4,3)
x = np.arange(n*c*h*w, dtype=np.float32).reshape(n, c, h, w)
np.random.shuffle(x)
x = 2*x/x.size -1
print "x="
print x


switch.enable_max_pooling = True
y1 = F.max_pooling_2d(x, 3, stride=2, pad=1, cover_all=False)
switch.enable_max_pooling = False
y2 = F.max_pooling_2d(x, 3, stride=2, pad=1, cover_all=False)

ydiff = y1-y2
print "ydiff(cover_all==False)="
print ydiff.data

switch.enable_max_pooling = True
y1 = F.max_pooling_2d(x, 3, stride=2, pad=1, cover_all=True)
switch.enable_max_pooling = False
y2 = F.max_pooling_2d(x, 3, stride=2, pad=1, cover_all=True)

ydiff = y1-y2
print "ydiff(cover_all==True)="
print ydiff.data
