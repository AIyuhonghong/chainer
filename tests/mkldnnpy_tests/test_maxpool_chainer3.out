x=
[[[[ 0.          0.02777779  0.05555558]
   [ 0.08333337  0.11111116  0.13888884]
   [ 0.16666663  0.19444442  0.22222221]
   [ 0.25        0.27777779  0.30555558]]

  [[ 0.33333337  0.36111116  0.38888884]
   [ 0.41666663  0.44444442  0.47222221]
   [ 0.5         0.52777779  0.55555558]
   [ 0.58333337  0.61111116  0.63888884]]

  [[ 0.66666663  0.69444442  0.72222221]
   [ 0.75        0.77777779  0.80555558]
   [ 0.83333337  0.86111116  0.88888884]
   [ 0.91666663  0.94444442  0.97222221]]]


 [[[-1.         -0.97222221 -0.94444442]
   [-0.91666669 -0.8888889  -0.8611111 ]
   [-0.83333331 -0.80555558 -0.77777779]
   [-0.75       -0.72222221 -0.69444442]]

  [[-0.66666663 -0.6388889  -0.6111111 ]
   [-0.58333337 -0.55555558 -0.52777779]
   [-0.5        -0.47222221 -0.44444442]
   [-0.41666669 -0.3888889  -0.3611111 ]]

  [[-0.33333331 -0.30555558 -0.27777779]
   [-0.25       -0.22222221 -0.19444442]
   [-0.16666669 -0.1388889  -0.1111111 ]
   [-0.08333331 -0.05555558 -0.02777779]]]]
ydiff(cover_all==False)=
[[[[ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]]]


 [[[ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]]]]
ydiff(cover_all==True)=
[[[[ 0.  0.]
   [ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]
   [ 0.  0.]]]


 [[[ 0.  0.]
   [ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]
   [ 0.  0.]]

  [[ 0.  0.]
   [ 0.  0.]
   [ 0.  0.]]]]
